### ISSUE TITLE
Naming Immutable Collections

### ISSUE THREAD

**Issue by GlenKPeterson** [Kotlin/kotlinx.collections.immutable#14]
Programmers are lazy.  Immutable collections are safer than mutable ones.  Therefore, they should have shorter names (less typing) than the mutable ones.  To this end I'd like to propose using "im" instead of "immutable" as a prefix for persistent/immutable collections and builders.  imListOf() instead of immutableListOf() makes it shorter than mutableListOf().  Still not as short as listOf().

I attended a Josh Bloch (I'll try to find the link later) where someone asked him about his biggest regret with the design of Java and he said it was that common collections weren't built into the language.  When I look at JSON in JavaScript, or the built-in collection syntax in Clojure, it's really sweet.  Arguably one of the stronger aspects of both languages that you can so briefly define data right in code.  To that end, I designed all my builders to just be 3 letters: vec(), set(), map(), and tup() for Vector, HashSet, HashMap, and Tuple2/3.

Kotlin has already used [] as a shortcut for getElementAt(), so we aren't getting Clojure/JavaScript -like syntax for immutable lists, and we're pretty much out of other symbols.  But I thought I'd throw it out there that really brief syntax for creating immutable data structures is awesome.  Especially if it's shorter than the syntax for creating mutable ones.  So whether you use "f" for frozen, "im" for immutable, or "p" for persistent, or something else, I just wanted to advocate for not having to type out the 9-letter word, "immutable" every time I reach for my preferred collections.

Still, we could follow Clojure's example of using # to prefix another symbol, so that #[1 2 3] could be a persistent/immutable list of ints.  #<1 2 3> might be a set.  #("one" 1) a tuple.  #{#("one" 1) #("two" 2)} a map.  It's a thought.  I put a lot of data structures in my code for little immutable things.  I don't like brevity in general, but for something like this, it's really sweet.

**Comment by ilya-g** [comment]
>  I just wanted to advocate for not having to type out the 9-letter word, "immutable" every time I reach for my preferred collections.

Even if we settle on `ImmutableList` in the end, you won't have to type it every time. Just type `ImList` and the completion will do the rest for you.

**Comment by stephenjfox** [comment]
What if, instead, the names were the same as the current Collections:

`kotlinx.collections.immutable.List` is exactly what you think it should be,
whereas `List` is what would be returned with `listOf()` (for example).

The imports (dis)ambiguate your code from the start, and the names are no longer than we're already used to.
Further, the IntelliSense can give us ALT+ENTER choices for imports

**Comment by cbruegg** [comment]
Having to choose between two imports every time I use collection-related
methods and classes sounds unconvenient, in my opinion.

Am Do., 5. Okt. 2017, 18:05 schrieb Stephen Fox <notifications@github.com>:

**Comment by jcornaz** [comment]
I would not bother too much about that. Users can create typealiases if they want shorter names in their code.

```kotlin
typealias ImList<E> = ImmutableList<E>
typealias PList<E> = PersistentList<E>
```

but from the library point of view, it's better to make it as clear as possible.

**Comment by Sporking** [comment]
Common use of typealiases for commonly used standard classes are a problem. If different groups of people create different aliases, then people won't be able to read each other's code. Also, what do you do if more than one alias for the same type is imported from the libraries you import? This can easily become a mess.

Also, having more than one List class also sounds painful to me. Java made this mistake with java.util.List and java.awt.List, and it was a persistent annoying problem for code that had to use both in the same class. For commonly used classes, this is painful.

I would like to see shorter names. After using Guava and having to deal with names like "ImmutableMultimap<String, ImmutableList<Integer>>", I think that some shortening would be a good thing. Four syllables in "immutable" is a bit too many in my opinion. I would be in favor of a prefix ("Im") or just a shorter word to use as a prefix. Some candidates that occur to me: "Fixed", "Hard", "Frozen". I think "Frozen" is perhaps my favorite, since it carries the meaning pretty well. There are lots of other options via https://www.thesaurus.com/browse/fixed,)

**Comment by jcornaz** [comment]
>If different groups of people create different aliases, then people won't be able to read each other's code. Also, what do you do if more than one alias for the same type is imported from the libraries you import? This can easily become a mess.

With any name it will become a mess if people don't use different types across the code base. For instance some people could use the types from the kotlin standard library, some may use immutable collection from guava (could be in the classpath for another reason), and some other would use `kotlinx.collections.immutable`.

Like almost everything when many people work together on the same code, it has to be a team decision, and people have to use the same idoms and naming scheme.

>Also, having more than one List class also sounds painful to me. Java made this mistake with java.util.List and java.awt.List, and it was a persistent annoying problem for code that had to use both in the same class. For commonly used classes, this is painful.

Fully agree. They have to have different names.

>Some candidates that occur to me: "Fixed", "Hard", "Frozen"

"Fixed" could be okay for me. But I really don't like "Hard" (what is a soft list?) and frozen is really not self-explanatory IMHO.

**Comment by fvasco** [comment]
Is `FixedList` a writable list with unmodifiable size, like the `Array`?

**Comment by jcornaz** [comment]
>Is FixedList a writable list with unmodifiable size, like the Array?

no... It would be an immutable persistent list. But thanks for the point, it en-light why `Fixed` is not a good naming either.

**Comment by popematt** [comment]
It's perhaps a gross abuse of the term `const`, but `ConstList`, `ConstSet`, etc. would imply that the elements cannot change.

We could also borrow from Rust and use [Cow](https://doc.rust-lang.org/std/borrow/enum.Cow.html). It's short and doesn't conflict with anything in the standard library (AFAIK). If you've heard of "copy on write" then it's a little bit familiar, and if you haven't heard of "copy on write", the name is strange enough that you probably need to look it up instead of making assumptions about its meaning (like you might with `FixedList`). 

For example, `CowList` could be an interface that has copy-on-write operations—maybe even including (naive) default implementations for them. `PersistentCowList` would implement that interface using persistence as it's particular strategy for how to handle the copy-on-write requirements.

**Comment by fvasco** [comment]
@popematt 
`COW` (Copy-On-Write) is an implementation detail; it does not describe the behavior.
For example, the [Java implementation](https://docs.oracle.com/en/java/javase/21/docs/api/java.base/java/util/concurrent/CopyOnWriteArrayList.html) of CopyOnWriteArrayList is mutable

**Comment by popematt** [comment]
@fvasco 
That's a good point.

The connection in my mind was that an immutable data structure must be copied to obtain a modified version of it, and so something like `add(element: E): CowList` would be copy-on-write from a _user_ perspective. (I.e. you perform a "mutating" operation and get a new copy with the changes applied, regardless of whether it's a complete copy or a persistent "copy".)

I don't hold a strong opinion about this—just looking through some issues as I've recently been diving into the details of this lib and thought I'd throw some ideas in.

**Comment by Laxystem** [comment]
I believe it shouldn't be made shorter; Compare:

```kotlin
interface Registry<T> : ImmutableMap<Identifier, T>
// vs
interface Registry<T> : ImMap<Identifier, T>
```

**Comment by DmitryNekrasov** [comment]
The canonical builders are the descriptive `persistentListOf`/`persistentSetOf`/`persistentMapOf`, matching the stdlib convention (`mutableListOf`, `buildList`); a short `im`/`p` prefix trades readability for keystrokes, which we'd rather not. Closing as wontfix.
